File: app/src/main/java/com/onemoresecret/msg_fragment_plugins/MsgPluginEncryptedMessage.java
package com.onemoresecret.msg_fragment_plugins;

import androidx.biometric.BiometricPrompt;
import androidx.fragment.app.Fragment;

import com.onemoresecret.HiddenTextFragment;
import com.onemoresecret.MessageFragment;
import com.onemoresecret.OutputFragment;
import com.onemoresecret.R;

import java.io.IOException;
import java.security.KeyStoreException;


public class MsgPluginEncryptedMessage extends MessageFragmentPlugin<byte[]> {
    private byte[] messageData;

    public MsgPluginEncryptedMessage(MessageFragment messageFragment,
                                     byte[] messageData, int applicationId) throws Exception {

        super(messageFragment, messageData, applicationId);
    }

    @Override
    protected void init(byte[] messageData, int applicationId) throws IOException {
        this.messageData = messageData;
    }

    @Override
    public void showBiometricPromptForDecryption() throws KeyStoreException {
        //nothing to decrypt
    }

    @Override
    public Fragment getMessageView() {
        if (messageView == null) {
            messageView = new HiddenTextFragment();
            context.getMainExecutor().execute(() -> {
                var message = new String(messageData);
                var hiddenTextFragment = (HiddenTextFragment) messageView;
                messageFragment.getHiddenState().observe(messageView,
                        hidden -> hiddenTextFragment.setText(hidden ? context.getString(R.string.hidden_text) : message));
                ((OutputFragment) outputView).setMessage(message, context.getString(R.string.decrypted_message));
            });
        }
        return messageView;
    }

    @Override
    public void onAuthenticationSucceeded(BiometricPrompt.AuthenticationResult result) {
        //no authentication
    }
}

File: app/src/main/java/com/onemoresecret/msg_fragment_plugins/MsgPluginEncryptedMessage.kt
package com.onemoresecret.msg_fragment_plugins

import androidx.biometric.BiometricPrompt
import androidx.compose.runtime.Composable
import androidx.fragment.app.FragmentActivity
import com.onemoresecret.R
import com.onemoresecret.composable.HiddenTextScreen

import com.onemoresecret.composable.OutputScreen
import com.onemoresecret.composable.OutputViewModel
import kotlinx.coroutines.flow.MutableStateFlow

open class MsgPluginEncryptedMessage(
    activity: FragmentActivity,
    protected val messageData: ByteArray?,
    hiddenState: MutableStateFlow<Boolean>,
    onNavigateBack: () -> Unit
) : MessageFragmentPlugin(activity, hiddenState, onNavigateBack) {

    private val message = String(messageData!!)
    private val outputViewModel = OutputViewModel(preferences)

    init {
        outputViewModel.setMessage(message, context.getString(R.string.decrypted_message))
    }

    override fun showBiometricPromptForDecryption() {
        //nothing to decrypt
    }

    @Composable
    override fun TopBarActions() {

    }

    @Composable
    override fun MessageView(hiddenState: Boolean) {
        HiddenTextScreen(text = if (hiddenState) context.getString(R.string.hidden_text) else message)
    }

    @Composable
    override fun OutputView() {
        OutputScreen(outputViewModel = outputViewModel)
    }

    override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
        //no authentication
    }
}
